Improve logging when cgroupfs mount fails

chkhalt opened the pull request:
When the variable `mounted` is false, we avoid to check the value of `cgroupInstance` because, after mount, it is related to an earlier state.

Fixes: #15997

brb reviewed (commented):
Thanks. Could you add to your commit msg `Fixes: e8ca92b45469 ("cgroups: Add stubs for non-Linux platforms")`. It will also help to check to which Cilium versions the fix has to be backported.

Also, cc @joestringer, who added the code, to double check whether we are not missing anything with the fix.

joestringer reviewed (changes requested):
Hi Joao, can you add some additional context to the commit message about how you hit a problem here? It's not obvious to me at a glance why this is a problem. What environment did you use, what Cilium configuration, what was the behaviour that was wrong?

FYI this goes back further than e8ca92b ("cgroups: Add stubs for non-Linux platforms") , that commit only moved code around. This bug likely goes way back to its initial introduction.

brb commented on pkg/cgroups/cgroups_linux.go:
```diff
@@ -58,9 +58,7 @@ func cgrpCheckOrMountLocation(cgroupRoot string) error {
-		if err := mountCgroup(); err != nil {
-			return err
-		}
+		return mountCgroup()
```
Why do we return even in the case of `err == nil`?

chkhalt replied:
It is to avoid checking `cgroupInstance`. When the cgroup is not yet mounted, this variable receives `false` as return of `mountinfo.IsMountFS`. Because of that, the function returns the error:

```C
if !cgroupInstance {
    return fmt.Errorf("Mount in the custom directory %s has a different filesystem than cgroup2", cgroupRoot)
}
```
As a consequence, the function `func getCgroupRootForKind() (string, error)` is never called. This is not a critical bug, but it causes cilium to use a cgroupv2 sub-hierarchy that is subtly different than expected.

brb replied:
Thanks for the explanation. I was not aware that not much happens in the function after we mount.

geyslan commented on pkg/cgroups/cgroups.go:
```diff
@@ -127,20 +127,19 @@ func CheckOrMountCgrpFS(mapRoot string, runsOnKind bool) {
-				} else {
-					setCgroupRoot(p)
-				}
+		if err := cgrpCheckOrMountLocation(mapRoot); err != nil {
```
I think this is the right way to check it and to raise the error as fatal. :+1:

joestringer replied:
In general, `Fatal()` should be avoided.

The current set of accepted uses for `Fatal()` in Cilium is fairly limited, mostly for when the user has explicitly configured a set of commandline options that is invalid.

As far as I can tell, we never required cgroups v2 before, and most environments do not have cgroups v2 enabled, and this function is unconditionally run. This seems dangerous, if someone upgrades Cilium to this new version it will cause Cilium to crash and force the user to somehow resolve the issue. Am I missing something?

chkhalt replied:
@joestringer, would it be better just notify the user that socket-lb will disabled and the reason (failure to mount cgroupv2)?

chkhalt replied:
I think it would be like this, right?:
https://github.com/cilium/cilium/blob/819942ea1cbdc4b88741ba0b27a36a83941129b8/pkg/cgroups/cgroups.go#L136-L140

aditighag replied:
We do need cgroupv2 fs mounted for attaching BPF cgroup programs in order for socket-lb to work. This has always been the case. So if kube-proxy replacement is enabled, I would suggest that we fail early in cilium agent.
I would defer the configuration to check to @brb. Should we gate the fatal log by kube-proxy-replacement config option or host-reachable services (probably latter)?

geyslan replied:
> As far as I can tell, we never required cgroups v2 before, and most environments do not have cgroups v2 enabled, and this function is unconditionally run. This seems dangerous, if someone upgrades Cilium to this new version it will cause Cilium to crash and force the user to somehow resolve the issue. Am I missing something?

I clearly understand what you mean, @joestringer, and I understand the concern, but I wonder what @brb thinks based on his [note](https://github.com/cilium/cilium/pull/15999#issuecomment-848609493):

> I think we should not ignore the error. The comment in the code says:
>        // Failed cgroup2 mount is not a fatal error, sockmap will be disabled however
> We cannot no longer assume this, as bpf_sock (aka socket-lb) also depends on the mount.

I know how simple changes like these can demand a deeper analysis so, please, let us know how to help in this matter.

Edit: @aditighag have already exposed more information. Tks.

joestringer replied:
Thanks for the context @aditighag , I wasn't thinking about the particular options that may require it.

I still think that Fatal is a bad idea. If we can perform a pure check of e.g. commandline args then I think it's reasonable to fatal out so that we can inform the user quickly and let them change the inputs they provide to Cilium. This will typically only occur upon first setup / install of Cilium. If a check involves the state of the system itself, then Cilium upgrades or system restart (could involve kernel upgrade) could then randomly trigger Cilium to stop running via the `Fatal()`. Cilium then is harder to debug since it's not running and we can't gather state from it during runtime, and it also increases the impact of issues like #14333. I'd say there should be a pretty small exclusive set of system state checks that could lead to `Fatal()`, primarily the baseline BPF availability check.

If there are options that require cgroup mounting and it cannot be mounted, I would suggest logging at Error level since Cilium is being asked to implement some functionality (eg socket LB) and it cannot implement that functionality, and furthermore it would not work around the issue. If there is a runtime workaround, then maybe issuing a Warning is sufficient.

aditighag replied:
@joestringer I have mixed feelings about this. If users have enabled socket-based functionality, and cilium is unable to provide that, how can we convey that more prominently (if socket-lb isn't effective, cilium datapath will fall back to per-packet lb?)? More importantly, if Cilium wasn't able to mount cgroup 2 fs, won't `init.sh` fail to attach the BPF cgroup programs anyway? That'll lead to agent pod crashing, no? This could be worse in terms of debugging.

joestringer replied:
It's a good question, but we really need to be careful that we don't make problems worse for users by crashing out the Cilium-agent pod when it could otherwise continue to operate without trouble. If we're not careful, we can cause more problems with the message than what the underling failure causes.

If socket LB is not effective but we're falling back to per-packet LB, then we are not breaking datapath functionality in general, we're gracefully degrading. We must not fatal out if we are gracefully degrading functionality. If the user isn't paying attention to the log messages and they're not noticing a problem, then is there even really a problem?

> More importantly, if Cilium wasn't able to mount cgroup 2 fs, won't init.sh fail to attach the BPF cgroup programs anyway? That'll lead to agent pod crashing, no? This could be worse in terms of debugging.

It looks like this is the case iff socket LB is enabled. Currently we mount cgroupv2 unconditionally. So one example that doesn't fit well here is if you deploy with socket LB / KPR disabled and happen to hit a cgroupv2 mount issue. In this situation, a warning message would allow Cilium to continue uninhibited. With a Fatal, we would break pod deployment, maybe even break L7 traffic forwarding (ref. #14333). Sure the user may notice more quickly but this can cause an outage rather than business as usual + investigate a weird log message later on.

In terms of debugging, while the final error message that states why the fatal occurred in this scenario would not include the failure to mount cgroupfs, it should still be clear just a couple of lines up in the logs so I think that's sufficient to still debug and understand the root cause pretty quickly.

aditighag replied:
>It looks like this is the case iff socket LB is enabled.

That's why I [suggested](https://github.com/cilium/cilium/pull/15999#discussion_r646973082) to only do a fatal log when kpr/host-reachable-svc is enabled. :) Regardless of fatal or not, we should log a warning only when these configs are enabled.

chkhalt replied:
Hi Folks, I update the code inserting a check for `EnableHostReachableServices == true`.
Now the warning will be printed only for this condition ;)

brb replied:
Keep in mind that the cgroupv2 is required for the sockops too. Also, if we fatal, then we need to check whether kpr != probe. As in the latter case, we can just disable the socket-lb.

geyslan commented on pkg/cgroups/cgroups.go:
```diff
@@ -127,20 +127,19 @@ func CheckOrMountCgrpFS(mapRoot string, runsOnKind bool) {
+		if runsOnKind {
+			p, err := getCgroupRootForKind()
+			if err != nil {
+				log.WithError(err).
```
Regarding the [conversation](https://github.com/cilium/cilium/pull/15999#issuecomment-848609493), should this be a Warning or a Fatal too?

geyslan commented on pkg/cgroups/cgroups_linux.go:
```diff
@@ -56,15 +56,12 @@ func cgrpCheckOrMountLocation(cgroupRoot string) error {
-
-	if !cgroupInstance {
+	// If the custom location is already mounted, check cgroupInstance, otherwise try to mount
+	if mounted && !cgroupInstance {
```
It seems to solve the problem of incorrectly testing `cgroupInstance` from a previous mount. Very good.

I would just like to ask if the cilium's way of checking things would agree with the commit or would it be something like this (curious about the code styling):

```
	// If the custom location is already mounted, check cgroupInstance, otherwise try to mount
	if mounted {
		if !cgroupInstance {
			return fmt.Errorf("Mount in the custom directory %s has a different filesystem than cgroup2", cgroupRoot)
		}
	} else {
		return mountCgroup()
	}
```

geyslan reviewed (approved):
Everything is ok. Just curious about `Fatal()` handling and some code styling.

joestringer reviewed (changes requested):
Removing my approval until we resolve the `Fatal()` handling.

aditighag commented on pkg/cgroups/cgroups.go:
```diff
@@ -127,20 +127,19 @@ func CheckOrMountCgrpFS(mapRoot string, runsOnKind bool) {
-					setCgroupRoot(p)
-				}
+		if err := cgrpCheckOrMountLocation(mapRoot); err != nil {
+			log.WithError(err).Fatal("Failed to mount cgroup2")
```
We need not fail early if socket-lb (the default kube-proxy disabled mode) is not enabled.
Also, we should probably mention why we have a fatal log. Something along these lines - "Failed to mount cgroup v2 filesystem needed by socket load balancing").

aditighag reviewed (commented):
Meant to request changes in the new revision.

Commit:
cgroups: Fix improper error on return

If cgroupv2 is not mounted, a logical mistake returns a error
even after a successful mount, which is silently ignored by
the caller function.

This commit fixes the logical mistake and takes care of print a warning
message if the cgroupv2 could not be mounted.

Fixes: #15997

Signed-off-by: Joao Victorino <joao@accuknox.com>

aditighag reviewed (changes requested):
Removing my approval until we've resolved the open question around how to handle error conditions.

joestringer commented on pkg/cgroups/cgroups.go:
```diff
@@ -52,9 +53,12 @@ func CheckOrMountCgrpFS(mapRoot string) {
+
+		if err := cgrpCheckOrMountLocation(mapRoot); err != nil {
+			if option.Config.EnableHostReachableServices {
+				log.WithError(err).Warnf("Failed to mount cgroupv2. Socket-based LB will not work.")
```
nit: No need to `Warnf`, since `Warnf` runs the logic to parse the string to find arguments and then handles the arguments. `Warn()` should be sufficient.

```suggestion
				log.WithError(err).Warn("Failed to mount cgroupv2. Socket-based LB will not work.")
```

joestringer reviewed (approved):
Just one minor nit but otherwise LGTM.

joestringer commented on pkg/cgroups/cgroups.go:
```diff
@@ -52,9 +53,12 @@ func CheckOrMountCgrpFS(mapRoot string) {
-		if err == nil {
+
+		if err := cgrpCheckOrMountLocation(mapRoot); err != nil {
+			if option.Config.EnableHostReachableServices {
```
Actually, given @brb 's most recent comment:

https://github.com/cilium/cilium/pull/15999#discussion_r663119101

Maybe it's better to just always warn. We _could_ go and list all of the different features that require cgroups here and check each option to decide whether to warn or not, but then any time that changes we need to update it. I think it'll be simpler maintenance to just warn, and if the user isn't using cgroups functionality it doesn't matter. If they are using cgroups functionality, then when we actually try to use cgroups it'll complain and there _will_ be log messages to help debug.
